File: core/geotime/src/main/java/org/locationtech/geowave/core/geotime/store/query/api/VectorAggregationQueryBuilder.java
package org.locationtech.geowave.core.geotime.store.query.api;

import org.locationtech.geowave.core.geotime.store.query.BaseVectorQueryBuilder;
import org.locationtech.geowave.core.geotime.store.query.VectorQueryConstraintsFactoryImpl;
import org.locationtech.geowave.core.geotime.store.query.aggregate.VectorAggregationQueryBuilderImpl;
import org.locationtech.geowave.core.index.persist.Persistable;
import org.locationtech.geowave.core.store.api.AggregationQuery;
import org.locationtech.geowave.core.store.api.AggregationQueryBuilder;
import org.opengis.feature.simple.SimpleFeature;

/**
 * An aggregation query builder particular for vector data. This should be
 * preferentially used to build AggregationQuery's for SimpleFeature data.
 *
 * @param <P>
 *            the type for input parameters
 * @param <R>
 *            the result type
 */
public interface VectorAggregationQueryBuilder<P extends Persistable, R> extends
		AggregationQueryBuilder<P, R, SimpleFeature, VectorAggregationQueryBuilder<P, R>>,
		BaseVectorQueryBuilder<R, AggregationQuery<P, R, SimpleFeature>, VectorAggregationQueryBuilder<P, R>>
{
	@Override
	default VectorQueryConstraintsFactory constraintsFactory() {
		return VectorQueryConstraintsFactoryImpl.SINGLETON_INSTANCE;
	}

	/**
	 * get a default implementation of this builder
	 *
	 * @return the builder
	 */
	static <P extends Persistable, R> VectorAggregationQueryBuilder<P, R> newBuilder() {
		return new VectorAggregationQueryBuilderImpl<>();
	}

	/**
	 * convenience method for getting a bounding box of the results of a query.
	 * It uses the default geometry for a feature type which is also the indexed
	 * geometry.
	 *
	 * @param typeNames
	 *            the type names to constrain by
	 * @return this builder
	 */
	VectorAggregationQueryBuilder<P, R> bboxOfResults(
			String... typeNames );

	/**
	 *
	 * convenience method for getting a bounding box of the results of a query
	 * this can be particularly useful if you want to calculate the bbox on a
	 * different field than the default/indexed Geometry
	 *
	 * @param typeName
	 *            the type name
	 * @param geomAttributeName
	 *            the geometry attribute name
	 * @return this builder
	 */
	VectorAggregationQueryBuilder<P, R> bboxOfResultsForGeometryField(
			String typeName,
			String geomAttributeName );

	/**
	 * convenience method for getting a time range of the results of a query.
	 * this has to use inferred or hinted temporal attribute names if uncertain
	 * what implicit time attributes are being used then explicitly set
	 * timeAttributeName in other method
	 *
	 * @param typeNames
	 *            the type names to constrain by
	 * @return this builder
	 */
	VectorAggregationQueryBuilder<P, R> timeRangeOfResults(
			String... typeNames );

	/**
	 *
	 * convenience method for getting a time range of the results of a query
	 * this can be particularly useful if you want to calculate the time range
	 * on a specific time field
	 *
	 * @param typeName
	 *            the type names to constrain by
	 * @param timeAttributeName
	 *            the time attribute name
	 * @return this builder
	 */
	VectorAggregationQueryBuilder<P, R> timeRangeOfResultsForTimeField(
			String typeName,
			String timeAttributeName );
}

File: python/src/main/python/pygw/query/vector/vector_aggregation_query_builder.py
#
# Copyright (c) 2013-2020 Contributors to the Eclipse Foundation

#
# See the NOTICE file distributed with this work for additional information regarding copyright
# ownership. All rights reserved. This program and the accompanying materials are made available
# under the terms of the Apache License, Version 2.0 which accompanies this distribution and is
# available at http://www.apache.org/licenses/LICENSE-2.0.txt
# ===============================================================================================

from pygw.config import geowave_pkg
from ...base.envelope import EnvelopeTransformer
from ...base.interval import IntervalTransformer

from ...base.type_conversions import StringArrayType
from ..aggregation_query_builder import AggregationQueryBuilder
from .vector_query_constraints_factory import VectorQueryConstraintsFactory


class VectorAggregationQueryBuilder(AggregationQueryBuilder):
    """
    A builder for creating aggregation queries for vector data.
    """

    def __init__(self):
        j_agg_qbuilder = geowave_pkg.core.geotime.store.query.api.VectorAggregationQueryBuilder.newBuilder()
        super().__init__(j_agg_qbuilder)

    def constraints_factory(self):
        """
        Creates a constraints factory for vector queries.  The vector query constraint factory
        provides additional constraints specific to vector data.

        Returns:
            A `pygw.query.vector.VectorQueryConstraintsFactory`.
        """
        return VectorQueryConstraintsFactory(self._java_ref.constraintsFactory())

    def bbox_of_results(self, *type_names):
        """
        Convenience method for getting a bounding box of the results of a query. It uses the default geometry for a
        feature type which is also the indexed geometry.

        Args:
            type_names (str): The type names to get the bounding box of.
        Returns:
            This query builder.
        """
        self._java_transformer = EnvelopeTransformer()
        if type_names is None:
            self._java_ref.bboxOfResults()
        else:
            self._java_ref.bboxOfResults(StringArrayType().to_java(type_names))
        return self

    def bbox_of_results_for_geometry_field(self, type_name, geometry_field_name):
        """
        Convenience method for getting a bounding box of the results of a query.  This can be particularly useful if you
        want to calculate the bbox on a different field than the default/indexed Geometry.

        Args:
            type_name (str): The type to aggregate.
            geometry_field_name (str): The geometry field to get the bounding box of.
        Returns:
            This query builder.
        """
        self._java_transformer = EnvelopeTransformer()
        self._java_ref.bboxOfResultsForGeometryField(type_name, geometry_field_name)
        return self

    def time_range_of_results(self, *type_names):
        """
        Convenience method for getting a time range of the results of a query.  This uses inferred or hinted
        temporal attribute names.

        Args:
            type_names (str): The type names to get the time range of.
        Returns:
            This query builder.
        """
        self._java_transformer = IntervalTransformer()
        self._java_ref.timeRangeOfResults(StringArrayType().to_java(type_names))
        return self

    def time_range_of_results_for_time_field(self, type_name, time_field_name):
        """
        Convenience method for getting a time range of the results of a query.  This can be particularly useful if you
        want to calculate the time range on a specific time field.

        Args:
            type_name (str): The type to aggregate.
            time_field_name (str): The time field to get the range of.
        Returns:
            This query builder.
        """
        self._java_transformer = IntervalTransformer()
        self._java_ref.timeRangeOfResultsForTimeField(type_name, time_field_name)
        return self

    def min(self, type_name, numeric_field_name):
        """
        Convenience method for getting the minimum value of a numeric field from the results of a query.

        Args:
            type_name (str): The type to aggregate.
            numeric_field_name (str): The numeric field to get the minimum value of.
        Returns:
            This query builder.
        """
        j_field_parameter = geowave_pkg.core.store.query.aggregate.FieldNameParam(numeric_field_name)
        j_min_agg = geowave_pkg.core.store.query.aggregate.FieldMinAggregation(j_field_parameter)
        return self.aggregate(type_name, j_min_agg)

    def max(self, type_name, numeric_field_name):
        """
        Convenience method for getting the maximum value of a numeric field from the results of a query.

        Args:
            type_name (str): The type to aggregate.
            numeric_field_name (str): The numeric field to get the maximum value of.
        Returns:
            This query builder.
        """
        j_field_parameter = geowave_pkg.core.store.query.aggregate.FieldNameParam(numeric_field_name)
        j_max_agg = geowave_pkg.core.store.query.aggregate.FieldMaxAggregation(j_field_parameter)
        return self.aggregate(type_name, j_max_agg)

    def sum(self, type_name, numeric_field_name):
        """
        Convenience method for getting the sum of a numeric field from the results of a query.

        Args:
            type_name (str): The type to aggregate.
            numeric_field_name (str): The numeric field to get the sum of.
        Returns:
            This query builder.
        """
        j_field_parameter = geowave_pkg.core.store.query.aggregate.FieldNameParam(numeric_field_name)
        j_sum_agg = geowave_pkg.core.store.query.aggregate.FieldSumAggregation(j_field_parameter)
        return self.aggregate(type_name, j_sum_agg)
